dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -158,12 +165,35 @@ public String toString() {
                     // the first call.
                     result = DialogueFutures.transform(
                             executeWithStickyToken(request, endpointChannel), successTransformer);
-                    result = DialogueFutures.catchingAllAsync(result, failureTransformer);
-                    callInFlight = Futures.nonCancellationPropagating(result);
+                    result = DialogueFutures.catchingAllAsync(result, failureTransformer, cancellationTransformer);
+                    callInFlight = result;
                 } else {
-                    result = callInFlightSnapshot;
-                    result = DialogueFutures.transformAsync(result, _input -> execute(request, endpointChannel));
-                    result = DialogueFutures.catchingAllAsync(result, _throwable -> execute(request, endpointChannel));
+                    // Each subsequent (parallel) call may be independently cancelled, that cancellation
+                    // must not leak to other pending calls.
+                    SettableFuture<Response> response = SettableFuture.create();
+                    DialogueFutures.addDirectListener(callInFlightSnapshot, () -> {
```

Comment:
This whole block is kinda scary, one wrong move and you're broken. Maybe trying to implement this queue as futures is not the way to go? I feel like at this point this needs to be a more specific implementation, like the QueuedChannel; you have a specific QueuedCall etc., with the SettableFuture as a field etc.

Whilst thinking about this I also realized that with this implementation, each of the requests are basically building a mini queue: if you're queued, you keep building a queue of futures, if the inFlightCall fails, you'll get notified, and attach yourself to the new inFlightCall; but the original future obviously needs to stay around, because that's what was returned to the caller. So you end up building a queue of linked futures, for each request and each failure.

Reply:
I think this queue of queues thing is maybe a problem, but I tried to pass down the SettableFuture at each step and it just gets messy, so probably not a worthwhile optimisation.

What I feel like we're missing is futures that have an explicit cleanup/resource attached to them, and then semantics on how the ownership of the resource is passed around. Kinda like Stream#close exists and also #onClose: feels like any sufficiently advanced flow data-structure should have a way of attaching resources that need closing/cleaning up.